File: bestia-core/src/main/java/net/bestia/core/game/model/Attack.java
package net.bestia.core.game.model;

import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.Id;

import net.bestia.core.game.battle.AttackBasedStatus;
import net.bestia.core.game.battle.Element;

@Entity
public class Attack {
	@Id
	private String databaseName;
	private int strength;
	private String name;
	@Enumerated(EnumType.STRING)
	private Element element;
	private int manaCost;
	@Enumerated(EnumType.STRING)
	private AttackBasedStatus basedStatus;
	
	public Attack() {
		// no op.
	}
	
	public Attack(String databaseName) {
		this.databaseName = databaseName;
	}
	
	/**
	 * @return the attackDbName
	 */
	public String getDatabaseName() {
		return databaseName;
	}
	
	/**
	 * @return the strength
	 */
	public int getStrength() {
		return strength;
	}
	
	/**
	 * @param strength the strength to set
	 */
	public void setStrength(int strength) {
		this.strength = strength;
	}
	
	/**
	 * @return the name
	 */
	public String getName() {
		return name;
	}
	
	/**
	 * @param name the name to set
	 */
	public void setName(String name) {
		this.name = name;
	}
	
	/**
	 * @return the element
	 */
	public Element getElement() {
		return element;
	}
	
	/**
	 * @param element the element to set
	 */
	public void setElement(Element element) {
		this.element = element;
	}
	
	/**
	 * @return the manaCost
	 */
	public int getManaCost() {
		return manaCost;
	}
	
	/**
	 * @param manaCost the manaCost to set
	 */
	public void setManaCost(int manaCost) {
		this.manaCost = manaCost;
	}
	
	/**
	 * @return the basedStatus
	 */
	public AttackBasedStatus getBasedStatus() {
		return basedStatus;
	}
	
	/**
	 * @param basedStatus the basedStatus to set
	 */
	public void setBasedStatus(AttackBasedStatus basedStatus) {
		this.basedStatus = basedStatus;
	}
}

File: bestia-models/src/main/kotlin/net/bestia/model/battle/Attack.kt
package net.bestia.model.battle

import java.io.Serializable
import javax.persistence.*

@Entity
@Table(name = "attacks")
class Attack(
    @Id
    val id: Int = 0,

    @Column(name = "attack_db_name", unique = true, nullable = false)
    val databaseName: String,

    val strength: Int,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val element: Element,

     val manaCost: Int = 0,

    /**
     * Flag tells if the attack has a script which needs to get executed upon
     * execution.
     */
    val hasScript: Boolean = false,

    /**
     * Range of the attack. Range is a mysql reserved word so an alias is needed.
     * Range is in meter.
     */
    @Column(name = "atkRange", nullable = false)
    val range: Int = 0,

    /**
     * Check if a line of sight to the target is necessairy.
     */
    val needsLineOfSight: Boolean = false,

    @Enumerated(EnumType.STRING)
    val type: AttackType,

    /**
     * Casttime in ms. 0 means it is instant.
     */
    val casttime: Int = 0,

    val cooldown: Int = 0,

    @Enumerated(EnumType.STRING)
    val target: AttackTarget
) : Serializable {

  val isRanged: Boolean
    get() = type == AttackType.RANGED_MAGIC || type == AttackType.RANGED_PHYSICAL

  /**
   * @return TRUE if the attack is magic or FALSE if its physical.
   */
  val isMagic: Boolean
    get() = type == AttackType.RANGED_MAGIC || type == AttackType.MELEE_MAGIC

  companion object {
    /**
     * Basic attack id used for the default attack every bestia has. Each bestia
     * has the default melee or ranged attack.
     */
    const val DEFAULT_MELEE_ATTACK_ID = -1
    const val DEFAULT_RANGE_ATTACK_ID = -2
    val DEFAULT_MELEE_ATTACK: Attack = Attack(
        id = DEFAULT_MELEE_ATTACK_ID,
        databaseName = "default_melee_attack",
        strength = 5,
        element = Element.NORMAL,
        manaCost = 0,
        range = 1,
        needsLineOfSight = true,
        type = AttackType.MELEE_PHYSICAL,
        target = AttackTarget.ENEMY_ENTITY,
        casttime = 0,
        cooldown = 1500
    )
  }
}
